### PR TITLE
 Gather attributes in the correct order when materializing

### PR DESCRIPTION
**Pull request by 2m**

### CODE REVIEW

**Comment by 2m** [comment]
PLS BUILD

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala [hunk 1]
```diff
@@ -3,70 +3,238 @@
...
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most outer dispatcher when specified directly around the graph stage" in {
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala]: innermost :), directly on the graph stage adds a more specific

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala [hunk 2]
```diff
@@ -3,70 +3,238 @@
...
+    "use the dispatcher set in materializer even if dispatcher specified in initial attributes" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)(materializerNoDispatcher)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "dispatcher in materializer trumps dispatcher specified directly around the graph stage" in {
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala]: This I disagree with, the materializer dispatcher should only be set if there was no other dispatcher attribute set.

File: akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala [hunk 3]
```diff
@@ -489,8 +489,9 @@ private final case class SavedIslandData(islandGlobalOffset: Int, lastVisitedOff
             matValueStack.addLast(result)
             if (Debug) println(s"COMP: $matValueStack")
           case PushAttributes(attr) ⇒
-            attributesStack.addLast(attributesStack.getLast and attr)
-            if (Debug) println(s"ATTR PUSH: $attr")
+            val mergedAttr = attr and attributesStack.getLast
+            if (Debug) println(s"ATTR PUSH: ${attributesStack.getLast} and $attr = $mergedAttr")
+            attributesStack.addLast(mergedAttr)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala]: I think it was right as it was.

**Review by johanandren — COMMENTED** [review]
We already discussed it but added the same things in comments here to clarify what I think.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala (hunk 2 above)
Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala]: +1

**Comment by johanandren** [comment]
This PR inverted the order which the materializer attributes are sandwiched into the attributes.

Closing this as we have decided to not go with any of the behavior changes for attributes.

**Pull request closed without merging**